### PR TITLE
Expanded browsercache options for 404 nginx handling #367

### PR DESCRIPTION
**Pull request by jacobd91**
#367

### CODE REVIEW

File: BrowserCache_ConfigLabels.php [hunk 1]
```diff
@@ -7,7 +7,11 @@ public function config_labels( $config_labels ) {
 				'browsercache.enabled' => __( 'Browser Cache:', 'w3-total-cache' ),
 				'browsercache.replace.exceptions' => __( 'Prevent caching exception list:', 'w3-total-cache' ),
 				'browsercache.no404wp' => __( 'Do not process 404 errors for static objects with WordPress', 'w3-total-cache' ),
-				'browsercache.no404wp.exceptions' => __( '404 error exception list:', 'w3-total-cache' ),
+				'browsercache.no404wp.response.cssjs' => __( '404 replacement response code for CSS/JS mime types', 'w3-total-cache' ),
+				'browsercache.no404wp.response.html' => __( '404 replacement response code for HTML mime types', 'w3-total-cache' ),
+				'browsercache.no404wp.response.other' => __( '404 replacement response code for Other mime types', 'w3-total-cache' ),
+				'browsercache.no404wp.response.exceptions' => __( '404 replacement response code for Exceptions URIs', 'w3-total-cache' ),
+				'browsercache.no404wp.exceptions' => __( '404 exception list:', 'w3-total-cache' ),
```
Comment (cssjoe) [inline BrowserCache_ConfigLabels.php:14]: You don't need to change old code if it's not relevant.

File: BrowserCache_Environment_Nginx.php [hunk 2]
```diff
@@ -131,12 +127,14 @@ public function generate( $mime_types, $cdnftp = false ) {
 
 			$impoloded = implode( '|', $exceptions );
 			if ( !empty( $impoloded ) ) {
-				$wp_uri = network_home_url( '', 'relative' );
-				$wp_uri = rtrim( $wp_uri, '/' );
+				$file_code = rtrim( network_home_url( '', 'relative' ), '/' ) . '/index.php?$args;';
+
+				if( $this->c->get_integer( 'browsercache.no404wp.response.exceptions' ) ){
```
Comment (cssjoe) [inline BrowserCache_Environment_Nginx.php:132]: WPCS: Add spaces for control structures. -- `if ( ... ) {`

File: BrowserCache_Environment_Nginx.php [hunk 3]
```diff
@@ -131,12 +127,14 @@ public function generate( $mime_types, $cdnftp = false ) {
...
+					$file_code = "=" . $this->c->get_integer( 'browsercache.no404wp.response.exceptions' );
```
Comment (cssjoe) [inline BrowserCache_Environment_Nginx.php:133]: Use single quotes for strings that do not need character translation or escaping other single quotes -- `'='`.

File: BrowserCache_Environment_Nginx.php [hunk 4]
```diff
@@ -302,19 +296,19 @@ private function generate_section( &$rules, $mime_types, $section ) {
 				$this->c, $section );
 			$rules .= '    ' . implode( "\n    ", $subrules ) . "\n";
 
-			if ( !$this->c->get_boolean( 'browsercache.no404wp' ) ) {
-				$wp_uri = network_home_url( '', 'relative' );
-				$wp_uri = rtrim( $wp_uri, '/' );
+			if ( $this->c->get_boolean( 'browsercache.no404wp' ) ) {
```
Comment (cssjoe) [inline BrowserCache_Environment_Nginx.php:299]: It looks like this condition was negated.  If `browsercache.no404wp` is true, then we should not be trying to rewrite the request through WordPress's `index.php` file.

**Review by cssjoe — CHANGES_REQUESTED** [review]
See file comments.  Also, the 404 handling affects Apache htaccess or Nginx conf.  The new features should be consistent with that.

File: BrowserCache_ConfigLabels.php (hunk 1 above)
Comment (jacobd91) [reply BrowserCache_ConfigLabels.php:14]: I don't follow. The only old code here I changed was the exceptions label for better consistency with the new options and a few other wording changes to hopefully avoid convusion

File: BrowserCache_Environment_Nginx.php (hunk 4 above)
Comment (jacobd91) [reply BrowserCache_Environment_Nginx.php:299]: I believe the original condition was actually incorrect as the generate_section is called for each mime type. If this condition were left as a !condition it wouldn't apply the "no 404" functionality. I will retest and confirm this

File: BrowserCache_ConfigLabels.php (hunk 1 above)
Comment (cssjoe) [reply BrowserCache_ConfigLabels.php:14]: It's the change from "404 error exception list" to "404 exception list".

**Review by maxicus — CHANGES_REQUESTED** [review]
extremely risky.

such change will break things with almost 100% probability.
many changes unrelated to issue itself.

i see at least 2 conditions "just removed"

**Comment by maxicus** [comment]
When no file present - it's passed to WP by default (that's what original WP rules do) and WP decides what to respond.
In a case of those blog his WP decided to return 301 - it's choice and it's how it should be. It's not W3TC's decision.
=> behavior described in original ticket is "expected and correct".
=> seems whole fix is just not needed.

**Comment by cssjoe** [comment]
Brad, Max, and I decided that this feature is not worth the risk of causing bigger issues by making more complex rewrite rules.

**Pull request closed without merging**